File: pushups.py
import cv2
import mediapipe as mp
import numpy as np

mp_drawing = mp.solutions.drawing_utils
mp_pose = mp.solutions.pose
mp_hands = mp.solutions.hands

pose = mp_pose.Pose(static_image_mode=False, model_complexity=1, min_detection_confidence=0.5, min_tracking_confidence=0.5)
hands = mp_hands.Hands(static_image_mode=False, max_num_hands=2, min_detection_confidence=0.5, min_tracking_confidence=0.5)

cap = cv2.VideoCapture(0)
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)

counter = 0
pushupsCount = 0
pushups = False
gesture_cooldown = 0
show_message = False
message_timer = 0
active = False

def calculate_angle(a, b, c):
    a = np.array([a.x, a.y])
    b = np.array([b.x, b.y])
    c = np.array([c.x, c.y])
    ba = a - b
    bc = c - b
    cosine_angle = np.dot(ba, bc) / (np.linalg.norm(ba) * np.linalg.norm(bc))
    angle = np.arccos(np.clip(cosine_angle, -1.0, 1.0))
    return np.degrees(angle)

while True:
    ret, frame = cap.read()
    if not ret:
        print("Camera read failed")
        break

    frame = cv2.flip(frame, 1)
    rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

    results_pose = pose.process(rgb)
    results_hands = hands.process(rgb)

    if results_hands.multi_hand_landmarks and results_pose.pose_landmarks and not active:
        nose = results_pose.pose_landmarks.landmark[0]
        for hand in results_hands.multi_hand_landmarks:
            wrist = hand.landmark[mp_hands.HandLandmark.WRIST]
            if abs(wrist.y - nose.y) < 0.1:
                active = True
                show_message = True
                message_timer = 60
                print(" Activated by hand on head")

    if active and results_pose.pose_landmarks:
        shoulderR = results_pose.pose_landmarks.landmark[11]
        elbowR = results_pose.pose_landmarks.landmark[13]
        wristR = results_pose.pose_landmarks.landmark[15]

        shoulderL = results_pose.pose_landmarks.landmark[12]
        elbowL = results_pose.pose_landmarks.landmark[14]
        wristL = results_pose.pose_landmarks.landmark[16]

        if counter % 3 == 0:
            angleR = calculate_angle(shoulderR, elbowR, wristR)
            angleL = calculate_angle(shoulderL, elbowL, wristL)

            if angleR < 95 and angleL < 95 and not pushups:
                pushupsCount += 1
                pushups = True
            elif angleR > 100 and angleL > 100:
                pushups = False

        counter += 1
        mp_drawing.draw_landmarks(frame, results_pose.pose_landmarks, mp_pose.POSE_CONNECTIONS)

    if results_hands.multi_hand_landmarks and gesture_cooldown == 0:
        if len(results_hands.multi_hand_landmarks) == 2:
            wrist1 = results_hands.multi_hand_landmarks[0].landmark[mp_hands.HandLandmark.WRIST]
            wrist2 = results_hands.multi_hand_landmarks[1].landmark[mp_hands.HandLandmark.WRIST]
            dx = abs(wrist1.x - wrist2.x)

            if dx > 0.30:
                pushupsCount = 0
                pushups = False
                active = False
                show_message = True
                message_timer = 60
                print(" Counter reset, waiting for activation")
                gesture_cooldown = 30

    if gesture_cooldown > 0:
        gesture_cooldown -= 1

    if active:
        cv2.putText(frame, f"Push-ups: {pushupsCount}", (20, 250), cv2.FONT_HERSHEY_PLAIN, 2, (0, 255, 0), 3)
    else:
        cv2.putText(frame, " Put hand on head to start", (20, 100), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 0), 2)

    if show_message:
        cv2.putText(frame, "🔁 Counter Reset" if not active else " Started", (20, 300), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2)
        message_timer -= 1
        if message_timer <= 0:
            show_message = False

    cv2.imshow("Push-up Counter with Gesture Control", frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

File: pushups.js
<!DOCTYPE html>
<html>
<head>
  <title>Push-Up Counter</title>
  <style>
    canvas { position: absolute; top: 0; left: 0; }
    #counter { position: absolute; top: 10px; left: 20px; font-size: 24px; color: lime; font-family: sans-serif; }
  </style>
</head>
<body>
  <video id="video" playsinline style="display:none;"></video>
  <canvas id="canvas"></canvas>
  <div id="counter">Push-ups: 0</div>

  <script type="module">
    import { Pose } from 'https://cdn.jsdelivr.net/npm/@mediapipe/pose/pose.js';
    import { Hands } from 'https://cdn.jsdelivr.net/npm/@mediapipe/hands/hands.js';
    import { Camera } from 'https://cdn.jsdelivr.net/npm/@mediapipe/camera_utils/camera_utils.js';
    import { drawConnectors, drawLandmarks } from 'https://cdn.jsdelivr.net/npm/@mediapipe/drawing_utils/drawing_utils.js';

    const video = document.getElementById('video');
    const canvas = document.getElementById('canvas');
    const ctx = canvas.getContext('2d');
    let pushups = 0;
    let active = false;
    let pushupState = false;
    let cooldown = 0;
    let showMessage = "";
    let messageTimer = 0;

    function calculateAngle(a, b, c) {
      const ab = [a.x - b.x, a.y - b.y];
      const cb = [c.x - b.x, c.y - b.y];
      const dot = ab[0]*cb[0] + ab[1]*cb[1];
      const magAB = Math.hypot(...ab);
      const magCB = Math.hypot(...cb);
      const cosine = dot / (magAB * magCB);
      const angle = Math.acos(Math.min(Math.max(cosine, -1.0), 1.0));
      return angle * 180 / Math.PI;
    }

    const pose = new Pose({
      locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/pose/${file}`
    });

    const hands = new Hands({
      locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/hands/${file}`
    });

    pose.setOptions({
      modelComplexity: 1,
      minDetectionConfidence: 0.5,
      minTrackingConfidence: 0.5
    });

    hands.setOptions({
      maxNumHands: 2,
      minDetectionConfidence: 0.5,
      minTrackingConfidence: 0.5
    });

    let poseResults = null;
    let handResults = null;

    pose.onResults(results => {
      poseResults = results;
    });

    hands.onResults(results => {
      handResults = results;
    });

    const camera = new Camera(video, {
      onFrame: async () => {
        await pose.send({ image: video });
        await hands.send({ image: video });
        draw();
      },
      width: 1280,
      height: 720
    });
    camera.start();

    function draw() {
      canvas.width = video.videoWidth;
      canvas.height = video.videoHeight;
      ctx.clearRect(0, 0, canvas.width, canvas.height);
      ctx.drawImage(video, 0, 0, canvas.width, canvas.height);

      if (poseResults && poseResults.poseLandmarks) {
        drawConnectors(ctx, poseResults.poseLandmarks, Pose.POSE_CONNECTIONS, { color: '#00FF00', lineWidth: 2 });
        drawLandmarks(ctx, poseResults.poseLandmarks, { color: '#FF0000', radius: 3 });

        const lShoulder = poseResults.poseLandmarks[11];
        const lElbow = poseResults.poseLandmarks[13];
        const lWrist = poseResults.poseLandmarks[15];
        const rShoulder = poseResults.poseLandmarks[12];
        const rElbow = poseResults.poseLandmarks[14];
        const rWrist = poseResults.poseLandmarks[16];
        const nose = poseResults.poseLandmarks[0];

        if (handResults && handResults.multiHandLandmarks && !active) {
          for (const hand of handResults.multiHandLandmarks) {
            const wrist = hand[0];
            if (Math.abs(wrist.y - nose.y) < 0.1) {
              active = true;
              pushups = 0;
              showMessage = "Started";
              messageTimer = 60;
              break;
            }
          }
        }

        if (active) {
          const angleL = calculateAngle(lShoulder, lElbow, lWrist);
          const angleR = calculateAngle(rShoulder, rElbow, rWrist);

          if (angleL < 95 && angleR < 95 && !pushupState) {
            pushups++;
            document.getElementById("counter").innerText = `Push-ups: ${pushups}`;
            pushupState = true;
          } else if (angleL > 100 && angleR > 100) {
            pushupState = false;
          }
        }
      }

      if (handResults && handResults.multiHandLandmarks && handResults.multiHandLandmarks.length === 2 && cooldown === 0) {
        const h1 = handResults.multiHandLandmarks[0][0];
        const h2 = handResults.multiHandLandmarks[1][0];
        if (Math.abs(h1.x - h2.x) > 0.3) {
          pushups = 0;
          document.getElementById("counter").innerText = `Push-ups: 0`;
          active = false;
          showMessage = "Counter Reset";
          messageTimer = 60;
          cooldown = 30;
        }
      }

      if (cooldown > 0) cooldown--;

      if (showMessage) {
        ctx.font = "30px Arial";
        ctx.fillStyle = "red";
        ctx.fillText(showMessage, 20, 350);
        messageTimer--;
        if (messageTimer <= 0) showMessage = "";
      }

      if (!active) {
        ctx.font = "24px Arial";
        ctx.fillStyle = "yellow";
        ctx.fillText("✋ Put hand on head to start", 20, 100);
      }

      requestAnimationFrame(draw);
    }
  </script>
</body>
</html>
